### PR TITLE
Implement user deprovisioning

### PR DESCRIPTION
**Pull request by jorissteyn**
This PR adds the deprovision API to EngineBlock. The API allows reading and deleting all user data for a specific user id:

 - GET /deprovision/{collab_person_id}
 - DELETE /deprovision/{collab_person_id}
 - DELETE /deprovision/{collab_person_id}/dry-run

The API is disabled by default, three INI settings are required to enable it:

```
engineApi.features.deprovision = 1
api.users.deprovision.username = "..."
api.users.deprovision.password = "..."
```

### CODE REVIEW

File: src/OpenConext/EngineBlock/Authentication/Model/User.php [hunk 1]
```diff
@@ -38,4 +39,17 @@ public function getCollabPersonUuid()
...
+
+    /**
+     * Specify data which should be serialized to JSON.
+     *
+     * @return array
+     */
+    public function jsonSerialize()
+    {
+        return [
+            'collab_person_id' => $this->getCollabPersonId()->getCollabPersonId(),
```
Comment (MKodde) [inline src/OpenConext/EngineBlock/Authentication/Model/User.php]: `getCollabPersonId()->getCollabPersonId()` looks kind of funny. Perhaps use the __toString feature instead?

File: src/OpenConext/EngineBlock/Service/DeprovisionService.php [hunk 2]
```diff
@@ -2,21 +2,155 @@
...
 namespace OpenConext\EngineBlock\Service;
 
+use OpenConext\EngineBlock\Authentication\Model\User;
+use OpenConext\EngineBlock\Authentication\Repository\ConsentRepository;
+use OpenConext\EngineBlock\Authentication\Repository\UserDirectory;
+use OpenConext\EngineBlock\Authentication\Value\CollabPersonId;
+use OpenConext\EngineBlockBundle\Authentication\Repository\SamlPersistentIdRepository;
+use OpenConext\EngineBlockBundle\Authentication\Repository\ServiceProviderUuidRepository;
+
 final class DeprovisionService implements DeprovisionServiceInterface
```
Comment (MKodde) [inline src/OpenConext/EngineBlock/Service/DeprovisionService.php]: This service takes quite a lot of responsibility:
- retrieve information
- format information
- remove information
- restructure response from a repository

Uncle Bob might not like this.

Suggestion, maybe introduce extra services, responsible for retrieving and removing the actual data. And maybe even add a service that is tasked with the formatting of the response data.

File: src/OpenConext/EngineBlockBundle/Authentication/Repository/ServiceProviderUuidRepository.php [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+     * @param string $uuid
+     * @return string|null
+     */
+    public function findEntityIdByUuid($uuid)
+    {
+        $entry = $this->findOneBy([
+            'uuid' => $uuid,
+        ]);
+
+        if ($entry !== null) {
```
Comment (MKodde) [inline src/OpenConext/EngineBlockBundle/Authentication/Repository/ServiceProviderUuidRepository.php]: why not test for `$entry instancof ServiceProvider` (or what the actual return type might be) ?

File: src/OpenConext/EngineBlockBundle/Controller/Api/DeprovisionController.php [hunk 4]
```diff
@@ -66,10 +68,12 @@ public function userDataAction(Request $request, $collabPersonId)
         $this->assertDeprovisionApiIsEnabled();
         $this->assertUserHasDeprovisionRole();
 
-        $userData = $this->deprovisionService->read($collabPersonId);
+        $id = $this->createCollabPersonIdValueObject($collabPersonId);
```
Comment (MKodde) [inline src/OpenConext/EngineBlockBundle/Controller/Api/DeprovisionController.php]: `$id` is not very descriptive. Why not just overwrite the `$collabPersonId` variable?

File: src/OpenConext/EngineBlock/Authentication/Model/User.php (hunk 1 above)
Comment (jorissteyn) [reply src/OpenConext/EngineBlock/Authentication/Model/User.php]: Can't do that, __toString() has a different purpose in EngineBlock; it formats for display.

File: src/OpenConext/EngineBlockBundle/Resources/config/controllers/api.yml [hunk 5]
```diff
@@ -13,6 +13,14 @@ services:
             - "@engineblock.features"
             - "@engineblock.service.consent"
 
+    engineblock.controller.api.deprovision:
+        class: OpenConext\EngineBlockBundle\Controller\Api\DeprovisionController
+        arguments:
+            - "@security.authorization_checker"
+            - "@engineblock.features"
+            - "@engineblock.service.deprovision"
+            - 'EngineBlock'
```
[the remark this answers is not in the document]
Comment (jorissteyn) [reply src/OpenConext/EngineBlockBundle/Resources/config/controllers/api.yml]: As discussed, adds complexity for something YAGNI.

**Pull request merged**